text/2024-01-27-autohinting.md

```diff
@@ -0,0 +1,266 @@
+information from the font to determine exactly *how* the outlines should be
+adjusted.
+
+This information falls into two categories:
+
+### Style classification ([src](https://github.com/googlefonts/fontations/blob/6eff17a15ca0caaea27eb130fa959d529d58625b/skrifa/src/outline/autohint/style.rs))
+
+Styles are effectively groups of glyphs that should be hinted in the same way
+with the same set of metrics (see the next section). These groups are broadly
+defined by Unicode script but may be further divided by specific OpenType
+features when the glyphs produced by those features would have different sizes
+and/or positions (e.g. subscripts, superscripts, small capitals).
+
+Style classification proceeds in three phases:
+1. Enumerate the character to glyph id mapping table and assign styles
+ roughly based on Unicode script. For example, the glyph id associated
+ with the character "A" would be assigned the "Latin default" style.
+2. Process the layout subsitution tables for known script/feature combinations
+ and assign the appropriate styles to glyphs that are *output* from those
+ substitutions. For example, any glyph id that is generated by the `smcp`
+ feature under the `latn` script would be assigned the "Latin small caps"
+ style.
+ 3. Process the layout substitution tables again, for each script, and assign
+ the default style for that script to any remaining unassigned glyph ids that
+ appear as outputs. This captures things like ligatures.
+
+This process is done for all glyphs at once when the autohinter state is
+initialized. The structure of a font file (lacking efficient mappings from
+glyph id to character or to layout feature) makes it cost prohibitive to do
+style classification for a single glyph.
+
+### Style metrics ([src](https://github.com/googlefonts/fontations/tree/6eff17a15ca0caaea27eb130fa959d529d58625b/skrifa/src/outline/autohint/metrics))
+
+Once we've classified all of our glyphs, the next step is to compute some
+metrics for each of the relevant styles in our glyph set. The purpose of these
+metrics is to ensure that glyph outlines of the same style are adjusted in a
+consistent manner to avoid jarring visual differences when the outlines are
+rendered together.
+
+Metrics consist of two related concepts:
+
+#### Alignment (aka "blue") zones ([src](https://github.com/googlefonts/fontations/blob/6eff17a15ca0caaea27eb130fa959d529d58625b/skrifa/src/outline/autohint/metrics/blues.rs))
+
+These define the extrema, or positions of outside edges for significant
+regions such as the top, bottom or x-height of a set of glyphs. Each style
+contains a list of candidate characters that are used to compute the
+positions of each alignment zone. For example, the maximum y value of the
+all points in the "T" glyph might be used to compute the top alignment
+zone while the minimum y value of all points in the "y" glyph could be
+use to compute the position of a bottom zone.
```

Comment:
Max y of T seems to suggest the zone is a line, do we expand this by some amount to create a 2d space (as "zone" seems to suggest)? The zone is anything within some distance of that line?